### PR TITLE
Supplies Screen

### PR DESCRIPTION
**Pull request by QuicheLorraine3600**
This PR adds the supplies screen to the detailed event page. From there, if you're a staff, you can add / edit / delte / assign supplies. As a guest, you can only assign / unassign yourself from a supply.

A new route (SUPPLIES_SCREEN) has been added.

Here are some screens:

- Staff perspective
image
image
image

- Guest perspective
image
image

### CODE REVIEW

**Comment by Athanaze** [comment]
According to the design system that was choosen, different types of text field should not be next to each other:

https://m2.material.io/components/text-fields#usage

2024-05-14-092048_1920x1080_scrot

"Unassigned myself" is a bit weird; here are some alternatives:

1. I withdrew from the task.
2. I removed myself from the assignment.
3. I relinquished my responsibilities for the project.
4. I stepped back from the role.
5. I excused myself from the duty.
6. I opted out of the assignment.
7. I recused myself from the task.
8. I disengaged from the responsibility.
9. I abstained from the project.
10. I released myself from the obligation.

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/GuestSupplyDialog.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      description = supply.description,
+      onDismissRequest = onDismissRequest,
+      modifier = Modifier.testTag("guest_supply_dialog"),
+      buttonDataList =
+          listOf(
+              ButtonData(
+                  stringResource(id = R.string.chimpagne_cancel),
+                  onClick = onDismissRequest,
+                  modifier = Modifier.testTag("guest_supply_cancel")),
+              if (supply.assignedTo.containsKey(loggedUserUID))
```
Comment (Athanaze) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/GuestSupplyDialog.kt:37]: if-else statements without brackets should not be used here, according to Kotlin's style guide :

https://github.com/Kotlin/kotlin-style-guide/issues/20

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/StaffSupplyDialog.kt [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        LazyColumn(modifier = Modifier.fillMaxWidth()) {
+          if (tempSupply.assignedTo.keys.isEmpty()) {
+            item { Text(stringResource(id = R.string.supplies_supply_not_assigned)) }
+          } else {
+            item {
+              Text(
+                  stringResource(id = R.string.supplies_supply_assigned_to),
+                  modifier = Modifier.testTag("staff_list"))
+            }
+            accounts.entries.forEach { (userUID, userAccount) ->
```
Comment (Athanaze) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/StaffSupplyDialog.kt:98]: A filter would be more approriate

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/StaffSupplyDialog.kt [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                        tempSupply = tempSupply.copy(assignedTo = tempSupply.assignedTo - userUID)
+                      }
+                }
+              }
+            }
+          }
+          if (!tempSupply.assignedTo.keys.containsAll(accounts.keys)) {
+            item { Text(stringResource(id = R.string.supplies_not_assigned_to)) }
+            accounts.entries.forEach { (userUID, userAccount) ->
+              if (tempSupply.assignedTo[userUID] != true) {
```
Comment (Athanaze) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/StaffSupplyDialog.kt:115]: != true should be changed to !tempSupply.assignedTo[userUID]

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/SuppliesScreen.kt [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                Icon(Icons.Default.Add, contentDescription = "Add")
+              }
+        }
+      },
+      topBar = {
+        TopAppBar(
+            title = { Text(stringResource(id = R.string.supplies_screen_title)) },
+            modifier = Modifier.shadow(4.dp),
+            navigationIcon = {
+              IconButton(onClick = { navObject.goBack() }) {
```
Comment (Athanaze) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/SuppliesScreen.kt:141]: The reusable component GoBackButton should be used here:

https://github.com/Chimpagne/ChimpagneApp/blob/main/app/src/main/java/com/monkeyteam/chimpagne/ui/components/Button.kt

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/SupplyCard.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      shape = RoundedCornerShape(16.dp),
+      colors = CardDefaults.cardColors(MaterialTheme.colorScheme.primary)) {
+        Row(
+            horizontalArrangement = Arrangement.SpaceBetween,
+            modifier = Modifier.fillMaxWidth().padding(16.dp)) {
+              Text(
+                  text = "${supply.quantity} ${supply.unit}",
+                  modifier = Modifier.testTag("supply_quantity_and_unit"))
+              Text(
+                  text = "${supply.assignedTo.keys.size} assigned",
```
Comment (Athanaze) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/SupplyCard.kt:36]: A resource string should be used here.

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/SupplyDialogAccountEntry.kt [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    loggedUserUID: ChimpagneAccountUID,
+    showCheckBox: Boolean = false,
+    checked: Boolean = false,
+    onCheck: (Boolean) -> Unit = {}
+) {
+  ListItem(
+      headlineContent = {
+        Text(
+            text =
+                "${account?.firstName} ${account?.lastName}" +
```
Comment (Athanaze) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/SupplyDialogAccountEntry.kt:29]: A bit hard to read

File: app/src/main/java/com/monkeyteam/chimpagne/ui/navigation/NavigationActions.kt [hunk 7]
```diff
@@ -16,6 +16,7 @@ object Route {
   const val MY_EVENTS_SCREEN = "myEvents"
   const val VIEW_DETAIL_EVENT_SCREEN = "viewDetailEventScreen"
   const val MANAGE_STAFF_SCREEN = "manageStaffScreen"
+  const val SUPPLIES_SCREEN = "SuppliesScreen"
```
Comment (Athanaze) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/navigation/NavigationActions.kt:19]: the previous values use a lowercase character as the first character

File: app/src/main/res/values-fr/strings.xml [hunk 8]
```diff
@@ -90,14 +91,32 @@
...
+    <!-- Chimpagne -->
+    <string name="chimpagne_cancel">Annuler</string>
+    <string name="chimpagne_edit">Éditer</string>
+    <string name="chimpagne_save">Sauvegarder</string>
+    <string name="chimpagne_confirm">Confirmer</string>
+    <string name="chimpagne_add">Ajouter</string>
+    <string name="chimpagne_delete">Supprimer</string>
+    <string name="chimpagne_you">Vous</string>
+
+    <!--    Supplies    -->
```
Comment (Athanaze) [inline app/src/main/res/values-fr/strings.xml:103]: Comments in resource strings are a very welcome addition.
Indeed, these files were getting too disorganized.

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/StaffSupplyDialog.kt (hunk 3 above)
Comment (QuicheLorraine3600) [reply app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/StaffSupplyDialog.kt:115]: tempSupply.assignedTo: Map<ChimpagneAccountUID, Boolean?> thus !tempSupply.assignedTo[userUID] doesn't work as tempSupply.assignedTo[userUID] can be null

File: app/src/main/java/com/monkeyteam/chimpagne/ui/navigation/NavigationActions.kt (hunk 7 above)
Comment (QuicheLorraine3600) [reply app/src/main/java/com/monkeyteam/chimpagne/ui/navigation/NavigationActions.kt:19]: It seems a bit anarchic 
image
A future PR will solve this issue.

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/SupplyDialogAccountEntry.kt (hunk 6 above)
Comment (QuicheLorraine3600) [reply app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/SupplyDialogAccountEntry.kt]: fixed

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/SuppliesScreen.kt (hunk 4 above)
Comment (QuicheLorraine3600) [reply app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/SuppliesScreen.kt]: amazing component, thx

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/StaffSupplyDialog.kt (hunk 2 above)
Comment (QuicheLorraine3600) [reply app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/StaffSupplyDialog.kt]: Done, :bahamas: would be proud

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/GuestSupplyDialog.kt (hunk 1 above)
Comment (QuicheLorraine3600) [reply app/src/main/java/com/monkeyteam/chimpagne/ui/event/details/supplies/GuestSupplyDialog.kt]: fixed

**Review by Athanaze — APPROVED** [review]
The required changes were made, this is ready to be merged

**Review by juaniac — CHANGES_REQUESTED** [review]
As stated by other reviewer this is good PR overall, the UI is great looking and the tests are very robust.

But I have to point out a huge oversight.
When a guest or staff member leaves the event, their assigned supplies are still stored in the database.
This shouldn't happen in my opinion, once a person from the event decides to leave, they  should automatically get removed from all supplies they were assigned to.

Thus this change should be addressed before merging.

**Comment by QuicheLorraine3600** [comment]
> As stated by other reviewer this is good PR overall, the UI is great looking and the tests are very robust.
> 
> But I have to point out a huge oversight. When a guest or staff member leaves the event, their assigned supplies are still stored in the database. This shouldn't happen in my opinion, once a person from the event decides to leave, they should automatically get removed from all supplies they were assigned to.
> 
> Thus this change should be addressed before merging.

This is a UI only PR... However I fixed this

**Review by Athanaze — APPROVED** [review]

**Pull request merged**